### PR TITLE
Handle keys for suppressed models consistently

### PR DESCRIPTION
**Pull request by dougbu**
- #7992

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs [hunk 1]
```diff
@@ -189,24 +189,26 @@ protected virtual bool Visit(ModelMetadata metadata, string key, object model)
...
-            key = entry?.Key ?? key ?? string.Empty;
-            metadata = entry?.Metadata ?? metadata;
-            var strategy = entry?.Strategy;
+            var nonNullKey = key ?? string.Empty;
 
+            // Use entry.Key even if null for SuppressValidation call because we might not have entries in model state.
             if (ModelState.HasReachedMaxErrors)
             {
-                SuppressValidation(key);
+                SuppressValidation(entry == null ? nonNullKey : entry.Key);
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs]: The issue kinda feels like the same sort of problem as - https://github.com/aspnet/Mvc/issues/7963. My proposed fix there was to not reset the validation on previously validated objects (See https://github.com/pranavkm/ModelStateOrderRepro/blob/e4cfaa33574a0496635d093191df863959b1a36e/ModelStateOrderRepro/Controllers/TestObjectModelValidator.cs#L57-L75). Would that work here?

Comment (dougbu) [reply src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs]: Didn't realize we already had a similar issue. This fix also works for #7963. Want a regression test for that one too?

I had previously considered your suggestion. But, it's problematic because `ModelStateDictionary` key collisions are too common and `!Metadata.ValidateChildren` suppressions (which have keys available) _should_ overwrite the status of complex type properties for consistency -- last parameter wins.

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs]: Sure, if this addresses both scenarios, that sounds great.

Comment (dougbu) [reply src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs]: I was wrong on both counts. My fix didn't address #7963 and `ValidationVisitor` doesn't overwrite `ModelValidationState.Invalid` elsewhere. I'll go with your approach and include the #7963 regression test (actually two test 😃).

## Details
Actually, `ValidationVisitor` will still overwrite `ModelValidationState.Invalid` in one corner case after taking your approach. Of course the following isn't relevant for either #7963 or #7992: https://github.com/aspnet/Mvc/blob/f12f9b46ed5cc848a36db1fa9933c35b3eb60351/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs#L103-L110

Comment (dougbu) [reply src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs]: Filed #8078 about that exception

**Comment by dougbu** [comment]
:up::date:

**Comment by dougbu** [comment]
f2608c2ff4
